### ISSUE TITLE
rewrite `...` to `..=` as an idiom lint for Rust 2018 edition

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#51043]
It might be nice to rewrite `...` to `..=` as an idiom lint when upgrading to the new edition.

**Comment by nikomatsakis** [comment]
Do we have an existing lint for this?

**Comment by zackmdavis** [comment]
If we hope to someday get rid of `...` in patterns (the current description of #28237 describes it as "(silently) deprecated"), wouldn't we want a future-incompatible lint rather than an idiom lint? (`...` in expressions is [already illegal](https://github.com/rust-lang/rust/blob/c8e10e386ab714c61479bc5c95ca6d42f2144143/src/libsyntax/parse/parser.rs#L4671-L4679).)

**Comment by kennytm** [comment]
One remaining issue of `a..=b` pattern is #48501. When stabilizing `a..=b`, to avoid the confusing interpretation of `&a...b`, in the following example:

```rust
fn main() {
    match &5 {
        &3...6 => {}
        _ => {}
    }
}
```

if we just `s/.../..=/g` the code will error with

```
error: the range pattern here has ambiguous interpretation
 --> src/main.rs:3:10
  |
3 |         &3..=6 => {}
  |          ^^^^^ help: add parentheses to clarify the precedence: `(3 ..=6)`
```

Note that `pattern_parentheses` (#48500) is still unstable 🙃. However, given that default binding modes (#42640) is already stable, an even more idiomatic fix would probably be

```rust
fn main() {
    match &5 {
        3..=6 => {}   // no &, no (), no whatever
        _ => {}
    }
}
```

**Comment by nikomatsakis** [comment]
@zackmdavis 

> If we hope to someday get rid of ... in patterns (the current description of #28237 describes it as "(silently) deprecated"), wouldn't we want a future-incompatible lint rather than an idiom lint? (... in expressions is already illegal.)

I classified it as an idiom lint because the code will not cause a **hard error** in the new edition, even if it is frowned upon.

**Comment by zackmdavis** [comment]
> if you've got a moment, can you write up a listing of what's left after #51149 landed?

@alexcrichton: the [comment above](https://github.com/rust-lang/rust/issues/51043#issuecomment-392252285) by @kennytm should still be addressed; the lint suggestion [is marked as maybe-incorrect](https://github.com/rust-lang/rust/blob/3244d53c790809872c558ec5a869ea2e1e770ac8/src/librustc_lint/builtin.rs#L1738-L1740) for this reason. If pattern-parentheses are stabilized (which itself reportedly [needs more testing](https://github.com/rust-lang/rust/issues/51087#issuecomment-397408820)), using them in for a correct suggestion when needed will be very easy.

**Comment by alexcrichton** [comment]
Ok great, thanks!

**Comment by pnkfelix** [comment]
visited for T-compiler triage. It seems like it *was* blocked on stabilization of the pattern-parentheses feature. So what are the next steps, and who wants to own this?

**Comment by pnkfelix** [comment]
more triage notes: P-high. @varkor has agreed to take this on and/or mentor it.

**Comment by nikomatsakis** [comment]
Removing from the milestone, as idiom lints are not part of the Rust 2018 release.
